### PR TITLE
DATA: Contributions

### LINKED ISSUE #6: DATA: Contribution
## Model: Contribution

| Field                  | Type                  | Description                                                |
|------------------------|-----------------------|------------------------------------------------------------|
| Id                     | Guid                  | Our internal identifier                                    |
| RepositoryId           | Guid                  | Foreign key to Repository.Id                               |
| ContributorId          | Guid                  | Foreign key to Contributor.Id                              |
| ContributionTypeId     | Guid                  | Foreign key to ContributionType.Id                         |
| ExternalId             | string                | Required, max 255 characters                              |
| Title                  | string                | Required, max 255 characters                              |
| CreatedAt              | DateTimeOffset       | The date/time the contribution / pull request was created  |
| UpdatedAt              | DateTimeOffset       | The date/time the contribution / pull request was updated  |
| MergedAt               | DateTimeOffset       | The date/time the contribution / pull request was merged    |

** **_Extra considerations: Add a unique constraint on (RepositoryId , ExternalId). This will ensure we do not have duplicate contributions for a repository._**

image

### PR DESCRIPTION
**Pull request by kandarpgautam**
closes #6 

- Added Contribution Model and migration.

### CODE REVIEW

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs:8]: Also rename the file to StorageBroker.Contributions.Configuration
```suggestion
        void AddContributionConfigurations(ModelBuilder modelBuilder)
```

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs:17]:
```suggestion
            modelBuilder.Entity<Contribution>()
                .HasIndex(contribution => new
                {
                    contribution.RepositoryId,
                    contribution.ExternalId
                })
                .IsUnique();
```

File: GitFyle.Core.Api/Models/Foundations/Contributions/Contribution.cs
Comment (cjdutoit) [inline GitFyle.Core.Api/Models/Foundations/Contributions/Contribution.cs:1]: add copyright

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs:10]: Do not make use of a variable to set the max length.  If you use the variable more than once it will lead to entanglement and you can unintendedly change other properties if you alter the value.  Rather do `.HasMaxLength(255)`

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs:22]:
```suggestion
            modelBuilder.Entity<Contribution>()
                .Property(contribution => contribution.ExternalId)
                .HasMaxLength(255)
                .IsRequired();
```

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs:27]:
```suggestion
            modelBuilder.Entity<Contribution>()
                .Property(contribution => contribution.Title)
                .HasMaxLength(255)
                .IsRequired();   
```

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.cs
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.cs:33]:
```suggestion
            AddContributionConfigurations(modelBuilder.Entity<Contribution>());
```

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs:30]:
```suggestion
            builder.HasIndex(contribution => new
                {
                    contribution.RepositoryId,
                    contribution.ExternalId
                })
                .IsUnique();

            builder.Property(contribution => contribution.ExternalId)
                .HasMaxLength(255)
                .IsRequired();

            builder.Property(contribution => contribution.Title)
                .HasMaxLength(255)
                .IsRequired();   
```

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs:14]: Add an IsRequired for properties:
RepositoryId
ContributorId
ContributionTypeId

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs:10]: Rename the file to `StorageBroker.Contributions.Configuration.cs`

Comment (kandarpgautam) [reply GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs:10]: should it be configuration or configurations?

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs:28]: Define the relationships between Contibution and Repository, Contributor, ContributionType

1) You will need to add `public IEnumerable<Contribution> Contributions { get; set; }` to the Repository.cs, Contributor.cs and ContributionType.cs
2) Add these to Contribution.cs

```cs
        public Repository Repository { get; set; }
        public Contributor Contributor { get; set; }
        public ContributionType ContributionType { get; set; }
```

3) Then in this configuration file you can define your one-to-many relationship like this:
```cs
            modelBuilder.Entity<Contribution>()
                .HasOne(contribution => contribution.Repository)
                .WithMany(respository => respository.Contributions)
                .HasForeignKey(contribution => contribution.RepositoryId)
                .OnDelete(DeleteBehavior.NoAction);

            modelBuilder.Entity<Contribution>()
                .HasOne(contribution => contribution.Contributor)
                .WithMany(contributor => contributor.Contributions)
                .HasForeignKey(contribution => contribution.ContributorId)
                .OnDelete(DeleteBehavior.NoAction);

            modelBuilder.Entity<Contribution>()
                .HasOne(contribution => contribution.ContributionType)
                .WithMany(contributionType => contributionType.Contributions)
                .HasForeignKey(contribution => contribution.ContributionTypeId)
                .OnDelete(DeleteBehavior.NoAction);
```
4) Add your migration changes

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs
Comment (cjdutoit) [reply GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs:10]: Configuration

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs
Comment (cjdutoit) [reply GitFyle.Core.Api/Brokers/Storages/StorageBroker.Contributions.References.cs]: > do I need to wait for Repository, Contributor, ContributionType to be merged?

If you pull in the latest code from main, then you will have access to `ContributionType`

Just follow the instructions above to complete the relationship configuration to `ContributionType` only.   Once `Contributor` and `Repository` is merged, we can do another `DATA` pull request to add relationship configuration for the the other two.

**Review by cjdutoit — CHANGES_REQUESTED** [review]
Resolve conflicts

**Comment by kandarpgautam** [comment]
> Resolve conflicts

Done.

**Review by cjdutoit — APPROVED** [review]
Fix

**Pull request merged**
